File: lab6_20_1/1.py
import heapq
import sys

def print_path(parent, goal):
    if parent[goal] == -1 and goal != -1: # Assuming start has parent -1
        # Check if goal is start
        pass
    
    path = []
    curr = goal
    while curr != -1:
        path.append(curr)
        curr = parent[curr]
    path.reverse()

    if path:
        print("Optimal Path : ")
        print(*(path))

def usc(adj, n, start, goal):
    # pq stores (cost, node)
    pq = [(0, start)]
    dist = [float('inf')] * (n + 1)
    parent = [-1] * (n + 1)
    dist[start] = 0
    
    while pq:
        cost, u = heapq.heappop(pq)

        if u == goal:
            print(f"min cost : {cost}")
            print_path(parent, goal)
            return

        if cost > dist[u]:
            continue
            
        if u in adj:
            for w, v in adj[u]:
                if dist[u] + w < dist[v]:
                    dist[v] = dist[u] + w
                    parent[v] = u
                    heapq.heappush(pq, (dist[v], v))

    print("Path not found")

def main():
    try:
        # For Uniform Cost Search input
        print("Enter the number of vertex and edges of the graph : ", end="", flush=True)
        line1 = sys.stdin.readline().split()
        if not line1: return
        n = int(line1[0])
        m = int(line1[1])
        
        print("Enter each edge's start node, end node and cost : ")
        adj = {}
        for _ in range(m):
            edge = sys.stdin.readline().split()
            if not edge: break
            u, v, c = map(int, edge)
            if u not in adj: adj[u] = []
            adj[u].append((c, v))

        print("Enter the start and goal node : ", end="", flush=True)
        line_sg = sys.stdin.readline().split()
        if not line_sg: return
        start = int(line_sg[0])
        goal = int(line_sg[1])
        
        usc(adj, n, start, goal)
        print("23052247\nKush Singh\nCSE-22")
    except (ValueError, EOFError, IndexError):
        pass

if __name__ == "__main__":
    main()

File: lab6_20_1/1.cpp
#include<bits/stdc++.h>
using namespace std;

void printPath(vector<int>&parent, int goal)
{
    if(parent[goal] == -1)
        return;
    
    vector<int>path;
    int curr = goal;
    while(curr != -1)
    {
        path.push_back(curr);
        curr = parent[curr]; 
    }
    reverse(path.begin(), path.end());

    cout << "Optimal Path : \n";
    for(int node : path)
        cout << node << " ";
    cout << "\n";
}

void USC(vector<pair<int,int>>adj[], int n, int start, int goal)
{
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq; // {cost, node}
    vector<int> dist(n+1, INT_MAX);
    vector<int> parent(n+1, -1);
    dist[start] = 0;
    pq.push({0, start});
    while(!pq.empty())
    {
        int cost = pq.top().first;
        int u = pq.top().second;
        pq.pop();

        if(u == goal)
        {
            cout << "min cost : " << cost << "\n";
            printPath(parent, goal);
            return;
        }

        if(cost > dist[u])
            continue;
        for(auto e : adj[u])
        {
            int w = e.first;
            int v = e.second;
            if(dist[u] + w < dist[v])
            {
                dist[v] = dist[u] + w;
                parent[v] = u;
                pq.push({dist[v], v});
            }
        }
    }

    cout << "Path not found\n";

}


int main()
{
    int n,m;
    cout << "Enter the number of vertex and edges of the graph : ";
    cin >> n >> m;
    cout << "Enter each edge's start node, end node and cost : \n";
    vector<pair<int,int>>adj[n+1];
    int u,v,c;
    for(int i = 0;i<m;i++)
    {
        cin >> u >> v >> c;
        adj[u].push_back({c,v});
    }

    int start, goal;
    cout << "Enter the start and goal node : ";
    cin >> start >> goal;
    USC(adj, n, start, goal);
    
    cout << "23052247\nKush Singh\nCSE-22\n";
    return 0;
}

/*
    Output :
    Enter the number of vertex and edges of the graph : 5 6
    Enter each edge's start node, end node and cost :
    1 2 2
    1 3 4
    2 3 1
    2 4 7
    3 5 3
    4 5 1
    Enter the start and goal node : 1 5
    min cost : 6
    Optimal Path :
    1 2 3 5
    23052247
    Kush Singh
    CSE-22
 */